eknrenderer/eknrenderer/eknr-renderer.c

```diff
@@ -0,0 +1,1408 @@
+    default:
+      return NULL;
+  }
+}
+
+static gboolean
+escape_value_eval_callback (const GMatchInfo *match_info,
+                            GString          *result,
+                            gpointer          user_data)
+{
+  /* Faster to just use if statements here */
+  g_autofree gchar *matched = g_match_info_fetch (match_info, 0);
+  char matched_ent = matched[0];
+  const gchar *html_entity = html_entity_for_char (matched_ent);
+
+  if (html_entity == NULL)
+    return TRUE;
+
+  g_string_append (result, html_entity);
+
+  /* From the documentation, FALSE to continue to the replacement process,
+   * TRUE to stop it. It isn't entirely clear, but it seems that prematurely
+   * stopping the replacement process is an error case. */
+  return FALSE;
+}
+
+/**
+ * eknr_escape_html:
+ * @html: The string to escape
+ * @error: A #GError
+ *
+ * Escape HTML, replacing characters with HTML entities.
+ *
+ * Returns: (transfer full): The escaped HTML document, or %NULL on error.
+ */
+gchar *
+eknr_escape_html (const gchar  *html,
+                  GError      **error)
+{
+  g_autoptr(GRegex) regex = NULL;
+  g_autofree gchar *escaped = NULL;
+
+  /* First, use a regex to pick out the relevant entities that
+   * need to be escaped.*/
+  regex = g_regex_new ("[&<>\"'\\/]", 0, 0, error);
+
+  if (regex == NULL)
+    return NULL;;
+
+  escaped = g_regex_replace_eval (regex,
+                                  html,
+                                  strlen (html),
+                                  0,
+                                  0,
+                                  escape_value_eval_callback,
+                                  NULL,
+                                  error);
+
+  if (escaped == NULL)
+    return NULL;
+
+  return g_steal_pointer (&escaped);
+}
+
+/**
+ * maybe_escape_value:
+ * @value: (transfer full): The string to escape
+ * @is_escaped: Whether or not to apply escaping
+ * @api: A #mustache_api_t, used for error reporting
+ *
+ * Based on what mustache.js does to escape HTML.
+ *
+ * Uses a regex to pick out offending characters, then replaces with
+ * HTML entities.
+ *
+ * Returns: %NULL on failure, an escaped value on success.
+ */
+static gchar *
+maybe_escape_value (gchar          *value,
+                    gboolean        is_escaped,
+                    mustache_api_t *api,
+                    void           *api_closure)
+{
+  g_autoptr(GError) error = NULL;
+  g_autofree gchar *unescaped = NULL;
+  g_autofree gchar *escaped = NULL;
+
+  if (!is_escaped)
+    return value;
+
+  /* Take an autofree reference to unescaped, it will be freed
+   * when the function returns. */
+  unescaped = value;
+  escaped = eknr_escape_html (unescaped, &error);
+
+  if (escaped == NULL)
+    {
+      (*api->error) (api, api_closure, __LINE__, error->message);
+      return NULL;
+    }
+
+  return g_steal_pointer (&escaped);
+}
+
+static uintmax_t
+_renderer_var_from_ht (mustache_api_t            *api,
```

Comment:
These `from_ht` functions should probably be renamed to `from_vardict`, since we're not using hash table anymore.